### ISSUE TITLE
Notebook with widgets don't run twice

### ISSUE THREAD

**Issue by remborg** [jupyterlab/jupyterlab#7144]
## Description

Notebooks with widgets can be ran only once. 

## Reproduce

1. Open a notebook with at least 1 component
2. Close it
3. Reopen it
4. The debugger should show 2 similar errors 'Uncaught (in promise) Error: Comms are disabled on this kernel connection'

I've been using a simple notebook, such as:
```
from ipywidgets import Button
btn = Button(description='test')
display(btn)
```

See 'Browser Output' below for screenshots

## Expected behavior

Opening a notebook after being closed should allow code cells to run.

## Context

- Operating System and version: Win 10
- Browser and version: Chrome 76.0.3809.132
- JupyterLab version: 1.1.1

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Library\mingw-w64\bin
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Library\usr\bin
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Library\bin
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Scripts
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\bin
        ...\AppData\Local\Continuum\anaconda3\condabin
        C:\WINDOWS\system32
        C:\WINDOWS
        C:\WINDOWS\System32\Wbem
        C:\WINDOWS\System32\WindowsPowerShell\v1.0
        C:\WINDOWS\System32\OpenSSH
        ...\AppData\Local\Microsoft\WindowsApps
        ...\AppData\Roaming\npm
        C:\Program Files\cmder\vendor\git-for-windows\cmd
        C:\Program Files\cmder\vendor\git-for-windows\mingw64\bin
        C:\Program Files\cmder\vendor\git-for-windows\usr\bin
        C:\Program Files\cmder

sys.path:
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Scripts
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\python37.zip
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\DLLs
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\lib
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\lib\site-packages

sys.executable:
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\python.exe

sys.version:
        3.7.3 (default, Apr 24 2019, 15:29:51) [MSC v.1915 64 bit (AMD64)]

platform.platform():
        Windows-10-10.0.17763-SP0

where jupyter:
        ...\AppData\Local\Continuum\anaconda3\envs\jupyterlab-ext\Scripts\jupyter.exe

pip list:
        Package             Version
        ------------------- ---------
        appdirs             1.4.3
        arrow               0.14.6
        asn1crypto          0.24.0
        astroid             2.2.5
        attrs               19.1.0
        backcall            0.1.0
        bifrostNotebooksApi 0.3
        binaryornot         0.4.4
        black               19.3b0
        bleach              3.1.0
        bokeh               1.2.0
        certifi             2019.6.16
        cffi                1.12.3
        chardet             3.0.4
        Click               7.0
        colorama            0.4.1
        cookiecutter        1.6.0
        cryptography        2.7
        decorator           4.4.0
        defusedxml          0.5.0
        entrypoints         0.3
        fileCrawler         0.5
        future              0.17.1
        idna                2.8
        ipykernel           5.1.2
        ipython             7.7.0
        ipython-genutils    0.2.0
        ipywidgets          7.5.0
        isort               4.3.21
        jedi                0.15.1
        Jinja2              2.10.1
        jinja2-time         0.2.0
        json5               0.8.5
        jsonschema          3.0.2
        jupyter-client      5.3.1
        jupyter-core        4.4.0
        jupyterlab          1.1.1
        jupyterlab-server   1.0.6
        lazy-object-proxy   1.4.1
        lxml                4.4.0
        MarkupSafe          1.1.1
        mccabe              0.6.1
        mistune             0.8.4
        nbconvert           5.6.0
        nbformat            4.4.0
        notebook            6.0.1
        numpy               1.17.1
        olefile             0.46
        packaging           19.0
        pandas              0.24.2
        pandocfilters       1.4.2
        parso               0.5.1
        pickleshare         0.7.5
        Pillow              6.1.0
        pip                 19.2.3
        poyo                0.5.0
        prometheus-client   0.7.1
        prompt-toolkit      2.0.9
        pycparser           2.19
        Pygments            2.4.2
        pylint              2.3.1
        pyOpenSSL           19.0.0
        pyparsing           2.4.2
        pyrsistent          0.15.4
        PySocks             1.7.0
        python-dateutil     2.8.0
        pytz                2019.2
        pywinpty            0.5.5
        PyYAML              5.1.2
        pyzmq               18.0.2
        requests            2.22.0
        Send2Trash          1.5.0
        setuptools          41.2.0
        six                 1.12.0
        terminado           0.8.2
        testpath            0.4.2
        toml                0.10.0
        tornado             6.0.3
        traitlets           4.3.2
        typed-ast           1.4.0
        urllib3             1.25.3
        wcwidth             0.1.7
        webencodings        0.5.1
        wheel               0.33.6
        whichcraft          0.6.0
        widgetsnbextension  3.5.1
        win-inet-pton       1.1.0
        wincertstore        0.2
        wrapt               1.11.2
</pre>
</details>

<details><summary>Browser Output</summary>
<pre>

![image](https://user-images.githubusercontent.com/9395250/64175711-c5580c00-ce53-11e9-9ab6-fb2c7fe6620b.png)

The error happens here:
![image](https://user-images.githubusercontent.com/9395250/64175753-e6206180-ce53-11e9-8cbf-332d8fbc466e.png)

</pre>
</details>

**Comment by jasongrout** [comment]
Thanks for reporting this. I can confirm this happens.

I've been thinking over the past day where the fundamental issue is. It appears it has to do with the document's context being closed when the document is closed, but the kernel is still left running and is around. When the document is opened again, it creates a new context, which creates a new kernel and a new widget manager. However, the new kernel notices that another kernel object is still around, so it defaults to not handling comms. And indeed it might - who knows what was listening to it and might be doing things.

I think either the context should stay around, or the context should clean up the kernel (which would possibly mean the kernel does not appear in the Running tab). Probably the context should stay around, which would mean things like js links would still work (since the widget manager and widget models would still be around).

**Comment by jasongrout** [comment]
Update on investigations:

It seems better for the context to not stay around, and it should clean up after itself properly. This does the trick:

```diff
diff --git a/packages/apputils/src/clientsession.tsx b/packages/apputils/src/clientsession.tsx
index c3a081b06..074cbe627 100644
--- a/packages/apputils/src/clientsession.tsx
+++ b/packages/apputils/src/clientsession.tsx
@@ -373,6 +373,7 @@ export class ClientSession implements IClientSession {
           console.error(`Kernel not shut down ${reason}`);
         });
       }
+      this._session.dispose();
       this._session = null;
     }
     if (this._dialog) {
```
It has the effect of when the client session disposes, the session is disposed, which in turn disposes the kernel connection. The problem with this is that the session and kernel connection objects, when they are disposed, emit terminated signals, and the kernel connection claims it is dead. Also the list of running sessions (the session models) is listening to the session terminated signal, and clears the session model from the list of running sessions when the session is disposed. This means the session disappears from the running tab and the green icon disappears next to the file name. When the sessions poller next runs, though, it realizes there is a session model from the server, and the running tab is updated and the green session icon reappears next to the file name.

Really, we should:
a. distinguish between models of what is on the server side from objects on the client side communicating with things
b. make it possible to dispose js objects without affecting our notion of what is on the server side
c. disentangle the notions of status of js objects from server-side objects

I started such work in the huge kernel architecture refactor which got postponed so we could ship 1.0.

**Comment by jasongrout** [comment]
Released in 1.1.2

**Comment by remborg** [comment]
This fixed the issue, thanks for patching this fast!
